libraries/AP_Motors/AP_MotorsHeli_Swash.cpp

```diff
@@ -86,130 +87,157 @@ const AP_Param::GroupInfo AP_MotorsHeli_Swash::var_info[] = {
+            // for 45 deg plates servos 1&2 are LF&RF, 3&7 are LR&RR.
+            add_servo_angle(CH_1,  -45.0, 1.0);
+            add_servo_angle(CH_2,   45.0, 1.0);
+            add_servo_angle(CH_3, -135.0, 1.0);
+            add_servo_angle(CH_4,  135.0, 1.0);
+            break;
     }
+
+}
+
+void AP_MotorsHeli_Swash::add_servo_angle(uint8_t num, float angle, float collective)
+{
+    add_servo_raw(num,
+                  cosf(radians(angle + 90)),
+                  cosf(radians(angle)),
+                  collective);
 }
 
-// get_servo_out - calculates servo output
-float AP_MotorsHeli_Swash::get_servo_out(int8_t ch_num, float pitch, float roll, float collective) const
+void AP_MotorsHeli_Swash::add_servo_raw(uint8_t num, float roll, float pitch, float collective)
+{
+    if (num >= _max_num_servos) {
+        // Indexing problem should never happen
+        return;
+    }
+
+    _enabled[num] = true;
+    _rollFactor[num] = roll * 0.45;
+    _pitchFactor[num] = pitch * 0.45;
+    _collectiveFactor[num] = collective;
+
+    // Setup output function
+    SRV_Channel::Aux_servo_function_t function = SRV_Channels::get_motor_function(_motor_num[num]);
+    SRV_Channels::set_aux_channel_default(function, _motor_num[num]);
+
+    // outputs are defined on a -500 to 500 range for swash servos
+    SRV_Channels::set_range(function, 1000);
+
+    // swash servos always use full endpoints as restricting them would lead to scaling errors
+    SRV_Channels::set_output_min_max(function, 1000, 2000);
+
+}
+
+// calculates servo output
+void AP_MotorsHeli_Swash::calculate(float roll, float pitch, float collective)
 {
     // Collective control direction. Swash moves up for negative collective pitch, down for positive collective pitch
     if (_collective_direction == COLLECTIVE_DIRECTION_REVERSED){
         collective = 1 - collective;
     }
 
-    float servo = ((_rollFactor[ch_num] * roll) + (_pitchFactor[ch_num] * pitch))*0.45f + _collectiveFactor[ch_num] * collective;
-    if (_swash_type == SWASHPLATE_TYPE_H1 && (ch_num == CH_1 || ch_num == CH_2)) {
-        servo += 0.5f;
-    }
+    for (uint8_t i = 0; i < _max_num_servos; i++) {
+        if (!_enabled[i]) {
+            // This servo is not enabled
+            continue;
+        }
 
-    // rescale from -1..1, so we can use the pwm calc that includes trim
-    servo = 2.0f * servo - 1.0f;
+        _output[i] = (_rollFactor[i] * roll) + (_pitchFactor[i] * pitch) + _collectiveFactor[i] * collective;
+        if (_swash_type == SWASHPLATE_TYPE_H1 && (i == CH_1 || i == CH_2)) {
+            _output[i] += 0.5f;
+        }
 
-    if (_make_servo_linear == 1) {
-        servo = get_linear_servo_output(servo);
-    }
+        // rescale from -1..1, so we can use the pwm calc that includes trim
+        _output[i] = 2.0f * _output[i] - 1.0f;
+
+        if (_make_servo_linear) {
+            _output[i] = get_linear_servo_output(_output[i]);
```

Comment:
There is no change to get_linear_servo_output here so not an issue now, but we should remember that this is another variable to the outputs that we are not testing in the motors test tool.